crates/shared/src/price_estimation/native_price_cache.rs

```diff
@@ -175,6 +190,50 @@ impl Inner {
     }
 }
 
+impl UpdateTask {
+    /// Single run of the background updating process.
+    async fn single_update(&self, inner: &Inner) {
+        let metrics = Metrics::get();
+        metrics
+            .native_price_cache_size
+            .set(inner.cache.lock().unwrap().len() as i64);
```

Comment:
One thing I missed earlier is that the `inner.cache` gets locked twice in quick succession with the introduction of the `native_price_cache_size` metric. However, since this only happens in the background task that doesn't get run very frequently making the code more ugly to save one lock is probably not worth it.
An easy "fix" would be to put the metric update into `sorted_tokens_to_update()` which wouldn't fit well with the function, though.

Reply:
Yes. I was considering another refactor later where the task holds the lock and passes the reference around to the functions using it. As you said this is just ugly but doesn't have a performance impact in practice so I wasn't sure about it.